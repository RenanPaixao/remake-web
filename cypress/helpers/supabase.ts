import { createClient } from '@supabase/supabase-js'

const e2eSupabase = createClient(import.meta.env.VITE_SUPABASE_URL, import.meta.env.VITE_SUPABASE_SERVICE_ROLE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
})

export { e2eSupabase }
