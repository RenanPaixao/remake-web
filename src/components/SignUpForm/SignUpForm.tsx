import { useFormik } from 'formik'
import { Button, ButtonGroup, Checkbox, FormControl, FormErrorMessage, FormLabel, Input } from '@chakra-ui/react'
import { useNavigate } from 'react-router-dom'
import { useState } from 'react'
import * as Yup from 'yup'
import { supabase } from '../../utils/supabase.ts'
import { useErrorToast } from '../../hooks/toast/useErrorToast.tsx'
import { hasFormikError } from '../../utils/hasFormikError.ts'

const validationSchema = Yup.object({
  firstName: Yup.string().required('Required').max(50, 'Name must be at most 50 characters'),
  lastName: Yup.string().required('Required').max(50, 'Last name must be at most 50 characters'),
  email: Yup.string().email('Invalid email address').required('Required').max(50, 'Email must be at most 50 characters'),
  password: Yup.string().required('Required').min(8, 'Password must be at least 8 characters'),
  isRecycler: Yup.boolean()
})

type FormValues = Yup.InferType<typeof validationSchema>

export const SignUpForm = () => {
  const errorToast = useErrorToast()
  const [isLoading, setIsLoading] = useState(false)

  const navigate = useNavigate()
  const formik = useFormik<FormValues>({
    initialValues: {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      isRecycler: false
    },
    onSubmit: async values => {
      setIsLoading(true)
      const { error } = await supabase.auth.signUp({
        email: values.email,
        password: values.password,
        options: {
          data: {
            first_name: values.firstName,
            last_name: values.lastName,
            is_recycler: values.isRecycler
          }
        }
      })
      setIsLoading(false)

      if(error) {
        errorToast({ description: error.message })
        console.error(error)
        return
      }

      navigate('/')
    },
    validationSchema,
    validateOnBlur: true
  })

  /**
   * Go to the login page.
   */
  function goToLogin() {
    navigate('/login')
  }

  return <form onSubmit={formik.handleSubmit}>
    <FormControl mt={4} isInvalid={hasFormikError('firstName', formik.touched, formik.errors)} isRequired>
      <FormLabel htmlFor='firstName'>First Name</FormLabel>
      <Input
        name='firstName'
        title='First Name'
        onChange={formik.handleChange}
        value={formik.values.firstName}
        onBlur={formik.handleBlur}
      />
      <FormErrorMessage>{formik.errors.firstName}</FormErrorMessage>
    </FormControl>

    <FormControl mt={4} isInvalid={hasFormikError('lastName', formik.touched, formik.errors)} isRequired>
      <FormLabel htmlFor='lastName'>Last Name</FormLabel>
      <Input
        name='lastName'
        title={'Last Name'}
        onChange={formik.handleChange}
        value={formik.values.lastName}
        onBlur={formik.handleBlur}
      />
      <FormErrorMessage>{formik.errors.lastName}</FormErrorMessage>
    </FormControl>

    <FormControl mt={4} isInvalid={hasFormikError('email', formik.touched, formik.errors)} isRequired>
      <FormLabel htmlFor='email'>Email</FormLabel>
      <Input
        name='email'
        title={'email'}
        onChange={formik.handleChange}
        value={formik.values.email}
        onBlur={formik.handleBlur}
        type={'email'}
      />
      <FormErrorMessage>{formik.errors.email}</FormErrorMessage>
    </FormControl>

    <FormControl mt={4} isInvalid={hasFormikError('password', formik.touched, formik.errors)} isRequired>
      <FormLabel htmlFor='password'>Password</FormLabel>
      <Input
        data-testid={'password-input'}
        name='password'
        title={'password'}
        type={'password'}
        onChange={formik.handleChange}
        value={formik.values.password}
        onBlur={formik.handleBlur}
      />
      <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
    </FormControl>

    <FormControl display={'flex'} mt={4} >
      <Checkbox
        name='isRecycler'
        onChange={formik.handleChange}
        isChecked={formik.values.isRecycler}
      >
        Are you a recycler?
      </Checkbox>
    </FormControl>

    <ButtonGroup mt={8} spacing={10} isDisabled={isLoading}>
      <Button variant={'outline'} isLoading={isLoading} onClick={goToLogin}>Login</Button>
      <Button
        colorScheme={'blue'}
        isLoading={isLoading}
        type={'submit'}
        isDisabled={formik.isSubmitting || Object.keys(formik.errors).length > 0}
      >
        Create Account
      </Button>
    </ButtonGroup>
  </form>
}
