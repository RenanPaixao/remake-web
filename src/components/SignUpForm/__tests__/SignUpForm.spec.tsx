import { fireEvent, screen, waitFor } from '@testing-library/dom'
import { act } from '@testing-library/react'
import { expect } from 'vitest'
import { customRender } from '../../../../tests/test-utils.tsx'
import { SignUpForm } from '../SignUpForm.tsx'
import { supabase } from '../../../utils/supabase.ts'

const supabaseSignUpMock = vi.fn()
supabase.auth.signUp = supabaseSignUpMock

const navigateMock = vi.fn()
const toastMock = vi.fn()

const formValues = {
  firstName: 'John',
  lastName: 'Doe',
  email: 'john.doe@email.com',
  password: '12345678',
  isRecycler: true
}

describe('SignUpForm', () => {
  beforeEach(() => {
    vi.resetAllMocks()
    vi.mock('react-router-dom', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useNavigate: () => navigateMock
      }
    })

    vi.mock('@chakra-ui/react', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useToast: () => toastMock
      }
    })
  })

  it('should render', async () => {
    const { container } = customRender(<SignUpForm />)

    expect(container).toMatchSnapshot()
  })

  it('should create an account successfully', async () => {
    supabaseSignUpMock.mockReturnValueOnce({ error: null })

    const { getByRole } = customRender(<SignUpForm/>)
    const submitButton = getByRole('button', { name: /create account/i })

    await fillForm(formValues)

    await act(async () => {
      fireEvent.click(submitButton)
    })

    await waitFor(() => {
      expect(supabaseSignUpMock).toHaveBeenCalledTimes(1)
      expect(supabaseSignUpMock.mock.calls[0][0]).toEqual({
        email: formValues.email,
        password: formValues.password,
        options: {
          data: {
            first_name: formValues.firstName,
            last_name: formValues.lastName,
            is_recycler: formValues.isRecycler
          }
        }
      })

      expect(navigateMock).toHaveBeenCalledTimes(1)
      expect(navigateMock.mock.calls[0][0]).toEqual('/')
    })
  })
  it('should not create an account successfully', async () => {
    const error = new Error('User already exists')
    supabaseSignUpMock.mockReturnValueOnce({ error  })

    const { getByRole } = customRender(<SignUpForm/>)
    const submitButton = getByRole('button', { name: /create account/i })

    await fillForm(formValues)

    await act(async () => {
      fireEvent.click(submitButton)
    })

    expect(toastMock.mock.calls[0][0]).toEqual(expect.objectContaining({
      description: error.message,
      status: 'error'
    }))
    expect(navigateMock).not.toBeCalled()
  })
})

const fillForm = async (values: Omit<typeof formValues, 'isRecycler'>) => {
  const {
    firstName,
    lastName,
    password,
    email
  } = values
  const { getByRole, getByTestId } = screen
  const firstNameInput = getByRole('textbox', { name: /first name/i })
  const lastNameInput = getByRole('textbox', { name: /last name/i })
  const emailInput = getByRole('textbox', { name: /email/i })
  // Password input has no role, so we need use the data-testid attribute
  const passwordInput = getByTestId('password-input')
  const isRecyclerInput = getByRole('checkbox')

  await act(async() => {
    fireEvent.change(firstNameInput, { target: { value: firstName } })
    fireEvent.change(lastNameInput, { target: { value: lastName } })
    fireEvent.change(emailInput, { target: { value: email } })
    fireEvent.change(passwordInput, { target: { value: password } })
    fireEvent.click(isRecyclerInput)
  })
}
