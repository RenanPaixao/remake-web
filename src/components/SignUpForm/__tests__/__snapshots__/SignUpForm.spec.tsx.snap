// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignUpForm > should render 1`] = `
<div>
  <form>
    <div
      class="chakra-form-control css-1mn9ptq"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="firstName"
        id="field-:r0:-label"
      >
        First Name
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-label="First Name"
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="field-:r0:"
        name="firstName"
        required=""
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-1mn9ptq"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="lastName"
        id="field-:r1:-label"
      >
        Last Name
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-label="Last Name"
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="field-:r1:"
        name="lastName"
        required=""
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-1mn9ptq"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="email"
        id="field-:r2:-label"
      >
        Email
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-label="email"
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        id="field-:r2:"
        name="email"
        required=""
        type="email"
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-1mn9ptq"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="password"
        id="field-:r3:-label"
      >
        Password
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-1tfjd1n"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-label="password"
        aria-required="true"
        class="chakra-input css-1cjy4zv"
        data-testid="password-input"
        id="field-:r3:"
        name="password"
        required=""
        type="password"
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-1jqyrav"
      role="group"
    >
      <label
        class="chakra-checkbox css-1577qb8"
      >
        <input
          aria-disabled="false"
          aria-invalid="false"
          class="chakra-checkbox__input"
          id="field-:r4:"
          name="isRecycler"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value=""
        />
        <span
          aria-hidden="true"
          class="chakra-checkbox__control css-1ydjfm6"
        />
        <span
          class="chakra-checkbox__label css-6x44c9"
        >
          Are you a recycler?
        </span>
      </label>
    </div>
    <div
      class="chakra-button__group css-1p2xg5l"
      data-orientation="horizontal"
      role="group"
    >
      <button
        class="chakra-button css-15jenze"
        type="button"
      >
        Login
      </button>
      <button
        class="chakra-button css-1tduvxh"
        type="submit"
      >
        Create Account
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
